packages/serverpod/lib/src/database/migrations/migration_manager.dart

```diff
@@ -141,42 +141,85 @@ class MigrationManager {
-        session,
-        where: (t) =>
-            t.module.equals(MigrationConstants.repairMigrationModuleName));
+    String? appliedVersionName = repairMigration.versionName;
+    await _withMigrationLock(session, () async {
+      var appliedRepairMigration = await DatabaseMigrationVersion.db
+          .findFirstRow(session,
+              where: (t) => t.module
+                  .equals(MigrationConstants.repairMigrationModuleName));
+
+      if (appliedRepairMigration != null &&
+          appliedRepairMigration.version == repairMigration.versionName) {
+        appliedVersionName = null;
+        return;
+      }
 
-    if (appliedRepairMigration != null &&
-        appliedRepairMigration.version == repairMigration.versionName) {
-      return null;
-    }
+      await session.db.unsafeSimpleExecute(
+        repairMigration.sqlMigration,
+      );
 
-    await session.db.unsafeSimpleExecute(
-      repairMigration.sqlMigration,
-    );
-    return repairMigration.versionName;
+      await _updateState(session);
+    });
+
+    return appliedVersionName;
   }
 
   /// Migrates all modules to the latest version.
   ///
   /// Returns the migrations applied.
   /// Returns null if latest version was already installed.
   Future<List<String>?> migrateToLatest(Session session) async {
-    var latestVersion = getLatestVersion();
+    List<String>? migrationsApplied = [];
 
-    var moduleName = session.serverpod.serializationManager.getModuleName();
+    await _withMigrationLock(session, () async {
+      await _updateState(session);
+      var latestVersion = getLatestVersion();
 
-    if (isVersionInstalled(moduleName, latestVersion)) {
-      return null;
-    }
+      var moduleName = session.serverpod.serializationManager.getModuleName();
 
-    var installedVersion = getInstalledVersion(moduleName);
+      if (isVersionInstalled(moduleName, latestVersion)) {
+        migrationsApplied = null;
+        return;
+      }
 
-    return await _migrateToLatestModule(
-      session,
-      latestVersion: latestVersion,
-      fromVersion: installedVersion,
-    );
+      var installedVersion = getInstalledVersion(moduleName);
+
+      migrationsApplied = await _migrateToLatestModule(
+        session,
+        latestVersion: latestVersion,
+        fromVersion: installedVersion,
+      );
+      await _updateState(session);
+    });
+
+    return migrationsApplied;
+  }
+
+  Future<void> _withMigrationLock(
+    Session session,
+    Future<void> Function() action,
+  ) async {
+    const String lockName = 'serverpod_migration_lock';
+
+    /// Use a transaction to ensure that the advisory lock is retained
+    /// until the transaction is completed.
+    ///
+    /// This ensures that we are only running migrations once at a time.
+    await session.db.transaction((transaction) async {
+      await session.db.unsafeExecute(
+        "SELECT pg_advisory_lock(hashtext('$lockName'));",
+        transaction: transaction,
```

Comment:
See the comment below. Reading up on this, I don't think the transaction here has any effect.